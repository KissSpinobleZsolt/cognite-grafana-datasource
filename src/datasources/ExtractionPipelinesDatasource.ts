import { DataQueryRequest, DataQueryResponse } from '@grafana/data';
import _ from 'lodash';
import { CogniteQuery, HttpMethod } from '../types';
import { Connector } from '../connector';

export class ExtractionPipelinesDatasource {
  public constructor(private connector: Connector) {}

  private runQuery(options) {
    const { id } = options;
    return this.connector.fetchItems({
      path: '/extpipes/runs/list',
      method: HttpMethod.POST,
      data: { filter: { externalId: id } },
    });
  }
  async query(options: DataQueryRequest<CogniteQuery>): Promise<DataQueryResponse> {
    const data = await Promise.all(
      options.targets.map((target) => {
        return this.runQuery({
          refId: target.refId,
          ...target.extractionPipelinesQuery,
        });
      })
    );
    return {
      data,
    };
  }
  async getAllExtractionPipelines() {
    return this.connector.fetchItems({
      path: '/extpipes',
      method: HttpMethod.GET,
      data: undefined,
    });
  }
}
