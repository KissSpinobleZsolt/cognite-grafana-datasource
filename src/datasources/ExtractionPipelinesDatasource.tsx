import { DataQueryRequest, DataQueryResponse, TimeRange } from '@grafana/data';
import _ from 'lodash';
import { CogniteQuery, HttpMethod } from '../types';
import { Connector } from '../connector';

interface ExtractionPipelinesResponse {
  status?: string | boolean;
  createdTime?: TimeRange;
  message?: string;
  [x: string]: any;
}
// eslint-disable-next-line no-nested-ternary
const evalStatus = (text) => (text === 'success' ? 1 : text === 'failure' ? 0 : 2);
export class ExtractionPipelinesDatasource {
  public constructor(private connector: Connector) {}

  private runQuery(options) {
    const { id, numeric } = options;
    return this.connector
      .fetchItems({
        path: '/extpipes/runs/list',
        method: HttpMethod.POST,
        data: { filter: { externalId: id } },
      })
      .then((response): ExtractionPipelinesResponse[] =>
        numeric
          ? _.map(response, ({ status, createdTime, ...rest }) => ({
              ...rest,
              status: evalStatus(status),
              createdTime: new Date(createdTime),
              id: `${id}`,
            }))
          : response
      );
  }
  async query(options: DataQueryRequest<CogniteQuery>): Promise<DataQueryResponse> {
    const data = await Promise.all(
      options.targets.map((target) => {
        return this.runQuery({
          refId: target.refId,
          ...target.extractionPipelinesQuery,
        });
      })
    );
    return {
      data,
    };
  }
  async getAllExtractionPipelines() {
    return this.connector.fetchItems({
      path: '/extpipes',
      method: HttpMethod.GET,
      data: undefined,
    });
  }
}
