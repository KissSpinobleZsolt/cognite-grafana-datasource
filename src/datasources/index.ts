export { ExtractionPipelinesDatasource } from './ExtractionPipelinesDatasource';
export { RelationshipsDatasource } from './RelationshipsDatasource';
export { TemplatesDatasource } from './TemplatesDatasource';
