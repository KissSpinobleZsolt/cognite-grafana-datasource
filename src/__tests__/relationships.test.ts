import { createRelationshipsNode } from '../datasources/RelationshipsDatasource';

describe('Relationships createRelationshipsNode', () => {
  const refId = 'A';
  describe('with empty relationships list', () => {
    const relationshipsList = [];

    it('return nodes', () => {
      expect(createRelationshipsNode(relationshipsList, refId)).toHaveProperty('nodes');
    });

    it('return edges', () => {
      expect(createRelationshipsNode(relationshipsList, refId)).toHaveProperty('edges');
    });
    it('match snapshot', () => {
      expect(createRelationshipsNode(relationshipsList, refId)).toMatchSnapshot();
    });
  });
  describe('adds value when relationshipsList is present', () => {
    const relationshipsList = [
      {
        externalId: 'test',
        sourceExternalId: 'test-source',
        source: {
          externalId: 'test-source',
          name: 'Test Source',
          description: 'test source description',
          metadata: {},
        },
        targetExternalId: 'test-target',
        target: {
          name: 'Test  Target',
          metadata: {},
          description: 'Test Target descriptipn',
          externalId: 'test-target',
        },
        labels: [{ externalId: 'Test Label ExternalId' }],
      },
    ];
    it('match snapshot', () => {
      expect(createRelationshipsNode(relationshipsList, refId)).toMatchSnapshot();
    });
  });
});
