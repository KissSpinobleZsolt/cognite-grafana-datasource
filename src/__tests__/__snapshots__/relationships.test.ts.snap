// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Relationships createRelationshipsNode adds value when relationshipsList is present match snapshot 1`] = `
Object {
  "edges": Object {
    "fields": Array [
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "id",
        "type": "string",
        "values": Array [
          "test",
        ],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "source",
        "type": "string",
        "values": Array [
          "test-source",
        ],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "target",
        "type": "string",
        "values": Array [
          "test-target",
        ],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "mainStat",
        "type": "string",
        "values": Array [
          "Test Label ExternalId",
        ],
      },
    ],
    "meta": Object {
      "preferredVisualisationType": "nodeGraph",
    },
    "name": "edges",
    "refId": "A",
  },
  "nodes": Object {
    "fields": Array [
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "id",
        "type": "string",
        "values": Array [
          "test-source",
          "test-target",
        ],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "title",
        "type": "string",
        "values": Array [
          "test source description",
          "Test Target descriptipn",
        ],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "mainStat",
        "type": "string",
        "values": Array [
          "Test Source",
          "Test  Target",
        ],
      },
    ],
    "meta": Object {
      "preferredVisualisationType": "nodeGraph",
    },
    "name": "nodes",
    "refId": "A",
  },
}
`;

exports[`Relationships createRelationshipsNode with empty relationships list match snapshot 1`] = `
Object {
  "edges": Object {
    "fields": Array [
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "id",
        "type": "string",
        "values": Array [],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "source",
        "type": "string",
        "values": Array [],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "target",
        "type": "string",
        "values": Array [],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "mainStat",
        "type": "string",
        "values": Array [],
      },
    ],
    "meta": Object {
      "preferredVisualisationType": "nodeGraph",
    },
    "name": "edges",
    "refId": "A",
  },
  "nodes": Object {
    "fields": Array [
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "id",
        "type": "string",
        "values": Array [],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "title",
        "type": "string",
        "values": Array [],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "mainStat",
        "type": "string",
        "values": Array [],
      },
    ],
    "meta": Object {
      "preferredVisualisationType": "nodeGraph",
    },
    "name": "nodes",
    "refId": "A",
  },
}
`;
